pyro/distributions/lkj.py

```diff
@@ -0,0 +1,205 @@
+    to the scale_tril parameter of a multivariate distribution such as MultivariateNormal.
+
+    E.g., if theta is a (positive) vector of covariances with the same dimensionality
+    as this distribution, and Omega is sampled from this distribution,
+    scale_tril=torch.mm(torch.diag(sqrt(theta)), Omega)
+
+    Note that the `event_shape` of this distribution is `[d, d]`
+
+    Important note: When using this distribution with HMC/NUTS, it is important to
+    use a `step_size` such as 1e-4. If not, you are likely to experience LAPACK
+    errors regarding positive-definiteness.
+
+    :param int d: Dimensionality of the matrix
+    :param torch.Tensor eta: A single positive number parameterizing the distribution.
+    """
+    arg_constraints = {"eta": constraints.positive}
+    support = corr_cholesky_constraint
+    has_rsample = True
+
+    def __init__(self, d, eta, validate_args=None):
+        if not torch.is_tensor(eta):
+            raise ValueError("Eta must be a tensor")
+        if any(eta <= 0):
+            raise ValueError("eta must be > 0")
+        if eta.numel() != 1:
+            raise ValueError("eta must be a single number; for a larger batch size, call expand")
+        vector_size = (d * (d - 1)) // 2
+        alpha = eta.add(0.5 * (d - 1.0))
+
+        concentrations = eta.new_empty(vector_size,)
+        i = 0
+        for k in range(d-1):
+            alpha -= .5
+            for j in range(k+1, d):
+                concentrations[..., i] = alpha
+                i += 1
+        self._gen = Beta(concentrations, concentrations)
+        self.eta = eta
+        self._d = d
+        self._lkj_constant = None
+        super(CorrLCholeskyLKJPrior, self).__init__(torch.Size(), torch.Size((d, d)), validate_args=validate_args)
+
+    def rsample(self, sample_shape=torch.Size()):
+        y = self._gen.rsample(sample_shape=self.batch_shape + sample_shape).detach()
+        z = y.to(dtype=self.eta.dtype).mul(2).add(-1.0)
+        return _vector_to_l_cholesky(z)
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(CorrLCholeskyLKJPrior, _instance)
+        batch_shape = torch.Size(batch_shape)
+        new._gen = self._gen
+        new.eta = self.eta
+        new._d = self._d
+        new._lkj_constant = self._lkj_constant
+        super(CorrLCholeskyLKJPrior, new).__init__(batch_shape, self.event_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    def lkj_constant(self, eta, K):
+        if self._lkj_constant is not None:
+            return self._lkj_constant
+
+        Km1 = K - 1
+
+        constant = torch.lgamma(eta.add(0.5 * Km1)).mul(Km1)
+
+        k = torch.linspace(start=1, end=Km1, steps=Km1, dtype=eta.dtype, device=eta.device)
+        constant -= (k.mul(math.log(math.pi) * 0.5) + torch.lgamma(eta.add(0.5 * (Km1 - k)))).sum()
+
+        self._lkj_constant = constant
+        return constant
+
+    def log_prob(self, x):
```

Comment:
This will have some subtle shape bugs, which makes the distribution batching semantics somewhat inconsistent with the rest of pytorch.

 - When you score a sample from `CorrLCholeskyLKJPrior(3, torch.tensor([1.]))` for example, its shape will be `[1]` instead of `[]` (i.e. it should return a torch scalar) which is the correct shape for a distribution with empty `batch_shape`. 
 - Another error would be when the data to be scored has a smaller shape than the distribution's batch_shape. In that case, other distributions would simply expand the `log_prob`. e.g. `CorrLCholeskyLKJPrior(3, torch.tensor([1.])).expand([5]).log_prob(torch.ones(3, 3)).shape == (5,)`. 

To debug these kind of shape bugs, I strongly recommend adding a fixture to `tests/distributions/conftest.py` - even a single example like below - and running `test_distributions.py`. I think the fixes should be relatively small. e.g. you might have to expand `eta` by `batch_shape` in `log_prob`.

```python
    Fixture(pyro_dist=dist.CorrLCholeskyLKJPrior,
            examples=[
                {'d': 3, 'eta': [1.],
                 'test_data': torch.ones(5, 3, 3)},
            ]),

```

Reply:
Thank you for that! I apologize that this has been somewhat seriatim - I wasn't familiar with the pyro test/build process before, and the templates I was working from didn't cover this.  I will take care of this new set of issues over the next day or two.

Reply:
Please take your time! We value your contributing to Pyro, and we realize that there's a learning curve to becoming an effective contributor (specially since documentation mostly caters to users and not contributors), and we are all here to help. 😄

Reply:
I've implemented the fixture and resolved the first issue you raise (when the batch_shape is empty, log_prob returns a scalar; when the batch shape is non-empty, log_prob returns a tensor of batch_shape; when sample shape is non-empty, log_prob returns a tensor of sample_shape; when both batch_shape and sample_shape are non-empty, log_prob returns a tensor of batch_shape + sample_shape).  

I'm a bit confused by 

> * Another error would be when the data to be scored has a smaller shape than the distribution's batch_shape

I'm not sure what this is referring to?  Does this mean that `log_prob` was called with an object of arbitrary shape?  The current behavior is that as long as the last two dimensions of the passed object match the dimensionality of the distribution, it will return a `log_prob` with a shape matching the passed-shape, less the last two dimensions. 

What **should** be the behavior in that case?

Reply:
> What **should** be the behavior in that case?

The shape/broadcasting semantics of PyTorch Distributions (and TensorFlow Distributions) specify that the shape of the output of `.log_prob(x)` is the result of [broadcasting](https://pytorch.org/docs/master/notes/broadcasting.html) the batch shape of the input (as determined by the distribution's `event_shape`) with the distribution's `batch_shape`.

Imagine a distribution with `batch_shape == (3,1,4)` and `event_shape == (5,)`. Here are several examples to illustrate:
- Any input tensor `x` with `x.shape[-1:] != (5,)` would cause an error, because it does not match the distribution's `event_shape`
- Any input tensor `x` with `x.shape[-2] != 4` or `x.shape[-4] != 3` would cause an error, because it does not match the non-empty dimensions of the distribution's `batch_shape`.
- An input tensor `x` with `x.shape == (5,)` would produce `.log_prob(x).shape == (3,1,4)`.
- An input tensor `x` with `x.shape == (2,4,5)` would produce `.log_prob(x).shape == (3,2,4)`.
- An input tensor `x` with `x.shape == (2,3,1,4,5)` would produce `.log_prob(x).shape == (2,3,1,4)`

Reply:
Super-helpful - thank you!